packages/legend-query/src/components/QueryBuilderPropertySearchPanel.tsx

```diff
@@ -0,0 +1,641 @@
+import type { QueryBuilderState } from '../stores/QueryBuilderState.js';
+import {
+  QueryBuilderSubclassInfoTooltip,
+  renderPropertyTypeIcon,
+} from './QueryBuilderExplorerPanel.js';
+import { QueryBuilderPropertyInfoTooltip } from './QueryBuilderPropertyInfoTooltip.js';
+import { QUERY_BUILDER_TEST_ID } from './QueryBuilder_TestID.js';
+
+const prettyPropertyNameFromNodeId = (name: string): string => {
+  let propNameArray = name.split('.');
+  propNameArray = propNameArray.map((p) => prettyPropertyName(p));
+  let propName = '';
+  propNameArray.forEach((p) => {
+    propName = `${propName + p}/`;
+  });
+  propName = propName.slice(0, -1);
+  return propName;
+};
+
+const prettyPropertyNameForSubType = (name: string): string => {
+  let propNameArray = name.split('@');
+  propNameArray = propNameArray
+    .map((p) => p.replace(/.*::/, ''))
+    .filter((p) => p !== '')
+    .map((p) => prettyPropertyName(p));
+  let propName = '';
+  propNameArray.slice(0, -1).forEach((p) => {
+    propName = `${propName}(@${p})/`;
+  });
+  propNameArray = guaranteeNonNullable(
+    propNameArray[propNameArray.length - 1],
+  ).split('.');
+  propNameArray = propNameArray.map((p) => prettyPropertyName(p));
+  propName = `${propName}(@${propNameArray[0]})/`;
+  propNameArray.slice(1).forEach((p) => {
+    propName = `${propName + p}/`;
+  });
+  propName = propName.slice(0, -1);
+  return propName;
+};
+
+const prettyPropertyNameForSubTypeClass = (name: string): string => {
+  let propNameArray = name.split('@');
+  propNameArray = propNameArray
+    .map((p) => p.replace(/.*::/, ''))
+    .filter((p) => p !== '')
+    .map((p) => prettyPropertyName(p));
+  let propName = '';
+  propNameArray.forEach((p) => {
+    propName = `${propName}@${p}`;
+  });
+  return propName;
+};
+
+export const QueryBuilderTreeNodeViewer = observer(
+  (props: {
+    node: QueryBuilderExplorerTreeNodeData;
+    queryBuilderState: QueryBuilderState;
+    explorerState: QueryBuilderExplorerState;
+    level: number;
+    stepPaddingInRem: number;
+  }) => {
+    const { node, queryBuilderState, explorerState, level, stepPaddingInRem } =
+      props;
+    const propertySearchPanelState = explorerState.propertySearchPanelState;
+    const [{ isDragging }, dragConnector] = useDrag(
+      () => ({
+        type:
+          node instanceof QueryBuilderExplorerTreePropertyNodeData
+            ? node.type instanceof Enumeration
+              ? QUERY_BUILDER_EXPLORER_TREE_DND_TYPE.ENUM_PROPERTY
+              : node.type instanceof Class
+              ? QUERY_BUILDER_EXPLORER_TREE_DND_TYPE.CLASS_PROPERTY
+              : QUERY_BUILDER_EXPLORER_TREE_DND_TYPE.PRIMITIVE_PROPERTY
+            : QUERY_BUILDER_EXPLORER_TREE_DND_TYPE.ROOT,
+        item: (): { node?: QueryBuilderExplorerTreePropertyNodeData } =>
+          node instanceof QueryBuilderExplorerTreePropertyNodeData
+            ? { node }
+            : {},
+        collect: (monitor): { isDragging: boolean } => ({
```

Comment:
I think I understand why you do what you did in the `onDrop` of other panels now, but I think you can checkout this doc

https://react-dnd.github.io/react-dnd/docs/api/use-drag

Basically, you can use `end()` hook to check when dragging ends instead